zookeeper-server/src/main/java/org/apache/zookeeper/client/ZooKeeperBuilder.java

```diff
@@ -0,0 +1,207 @@
+     *            basically means that if the client can't find any majority
+     *            servers but there's partitioned server it could reach, it
+     *            connects to one in read-only mode, i.e. read requests are
+     *            allowed while write requests are not. It continues seeking for
+     *            majority in the background.
+     * @return this
+     * @since 3.4
+     */
+    public ZooKeeperBuilder withCanBeReadOnly(boolean canBeReadOnly) {
+        this.canBeReadOnly = canBeReadOnly;
+        return this;
+    }
+
+    /**
+     * Specifies session id and password in session reestablishment.
+     *
+     * @param sessionId
+     *            session id to use if reconnecting, otherwise 0 to open new session
+     * @param sessionPasswd
+     *            password for this session
+     * @return this
+     * @see ZooKeeper#getSessionId()
+     * @see ZooKeeper#getSessionPasswd()
+     */
+    @SuppressFBWarnings({"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
+    public ZooKeeperBuilder withSession(long sessionId, byte[] sessionPasswd) {
+        this.sessionId = sessionId;
+        this.sessionPasswd = sessionPasswd;
+        return this;
+    }
+
+    /**
+     * Specifies the client config used to construct ZooKeeper instances.
+     *
+     * @param clientConfig
+     *            passing this conf object gives each client the flexibility of
+     *            configuring properties differently compared to other instances
+     * @return this
+     * @since 3.5.2
+     */
+    public ZooKeeperBuilder withClientConfig(ZKClientConfig clientConfig) {
+        this.clientConfig = clientConfig;
+        return this;
+    }
+
+    /**
+     * Creates a {@link ZooKeeperOptions} with configured options.
+     */
+    public ZooKeeperOptions toOptions() {
+        return new ZooKeeperOptions(
+            connectString,
+            sessionTimeout,
+            defaultWatcher,
+            hostProvider,
+            canBeReadOnly,
+            sessionId,
+            sessionPasswd,
+            clientConfig
+        );
+    }
+
+    /**
+     * Constructs an instance of {@link ZooKeeper}.
+     *
+     * @return an instance of {@link ZooKeeper}
+     * @throws IOException from constructor of {@link ZooKeeper}
+     */
+    public ZooKeeper build() throws IOException {
+        return new ZooKeeper(toOptions());
+    }
+
+    /**
+     * Constructs ZooKeeper instance using constructor of given class.
+     *
+     * @param clazz class of target ZooKeeper instance
+     * @return ZooKeeper instance
+     * @param <T> type of ZooKeeper instance
+     * @throws IllegalArgumentException if given class does not export required constructor
+     * @throws RuntimeException from constructor of ZooKeeper instance
+     * @throws IOException from constructor of ZooKeeper instance or wrapper of no IO exception
+     */
+    @SuppressWarnings("unchecked")
+    public <T extends ZooKeeper> T build(Class<T> clazz) throws IOException {
```

Comment:
Please don't go this way.

This is totally out of our control.
If developers know the class they can instantiate it explicitly.

It it fine to add a  buildZooKeeperAdmin() method

Reply:
I felt similar(multiple ways to construct). I think there are candidates for us to go:

1. Only `build` and `buildAdmin` for `ZooKeeper` and `ZooKeeperAdmin`. For other derivations, they should resort to `ZooKeeperBuilder.toOptions` and `CustomZooKeeper(ZooKeeperOptions)` for full options customization.
2. Restrict `ZooKeeperBuilder.toOptions` to some level of private, so `CustomZooKeeper(ZooKeeperOptions)` is a simple hook for `ZooKeeperBuilder`. This way clients are encouraged(or forced) to use `ZooKeeperBuilder::build(Class<T> class)` to construct `CustomZooKeeper`.
3. Combine above two and support only `ZooKeeper` and `ZooKeeperAdmin`. No third party derivations are supported or encouraged.

Maybe we can go 3 and unleash restriction in future if requested ? Currently, all other derivations are test purpose.

For "some level of private", I means `@InterfaceAudience.Private` or moving to package where `ZooKeeper` resides in.

Reply:
I have added a follow up commit to build only `ZooKeeper` and `ZooKeeperAdmin`(e.g. candidate 3 from above).

I limited `ZooKeeper(ZooKeeperOptions options)`, `ZooKeeperAdmin(ZooKeeperOptions options)` and `ZooKeeperOptions` to private using `InterfaceAudience.Private` now. `ZooKeeperAdmin` resides in different package than `ZooKeeper`, so I have resorted to `InterfaceAudience.Private` somewhere.

Reply:
> No third party derivations are supported or encouraged

Anyone inherits `ZooKeeper` should be fine to build their own constructing method.